Foundation/NSURLResponse.swift

```diff
@@ -178,4 +189,136 @@ public class NSHTTPURLResponse : NSURLResponse {
     */
     public class func localizedStringForStatusCode(statusCode: Int) -> String { NSUnimplemented() }
 }
+/// Parses the expected content length from the headers.
+///
+/// Note that the message content length is different from the message
+/// transfer length.
+/// The transfer length can only be derived when the Transfer-Encoding is identity (default).
+/// For compressed content (Content-Encoding other than identity), there is not way to derive the
+/// content length from the transfer length.
+private func expectedContentLengthFromHeaderFields(headerFields: [String : String]?) -> Int64? {
+    guard
+        let f = headerFields,
+        let contentLengthS = valueForCaseInsensitiveKey("content-length", fields: f),
+        let contentLength = Int64(contentLengthS)
+        else { return nil }
+    let contentEncoding = valueForCaseInsensitiveKey("content-encoding", fields: f)
+    let transferEncoding = valueForCaseInsensitiveKey("transfer-encoding", fields: f)
+    guard (contentEncoding == nil) || (contentEncoding == "identity") else { return nil }
+    guard (transferEncoding == nil) || (transferEncoding == "identity") else { return nil }
+    return contentLength
+}
+/// Parses the suggested filename from the `Content-Disposition` header.
+///
+/// - SeeAlso: [RFC 2183](https://tools.ietf.org/html/rfc2183)
+private func suggestedFilenameFromHeaderFields(headerFields: [String : String]?) -> String? {
+    // Typical use looks like this:
+    //     Content-Disposition: attachment; filename="fname.ext"
+    guard
+        let f = headerFields,
+        let contentDisposition = valueForCaseInsensitiveKey("content-disposition", fields: f),
+        let parts = contentDisposition.httpHeaderParts
+        else { return nil }
+    for part in parts.1 where part.0 == "filename" {
+        return part.1?.lastPathComponent
+    }
+    return nil
+}
+/// Parses the `Content-Type` header field
+///
+/// - Returns: Either a media type and `nil`, or a mesia type and a character set.
+///
+/// `Content-Type: text/html; charset=ISO-8859-4` would result in `("text/html", "ISO-8859-4")`, while
+/// `Content-Type: text/html` would result in `("text/html", nil)`.
+private func mimeTypeAndTextEncodingFromHeaderFields(headerFields: [String : String]?) -> (String,String?)? {
+    guard
+        let f = headerFields,
+        let contentType = valueForCaseInsensitiveKey("content-type", fields: f),
+        let parts = contentType.httpHeaderParts
+        else { return nil }
+    for part in parts.1 where part.0 == "charset" {
+        return (parts.0, part.1)
+    }
+    return (parts.0, nil)
+}
 
+private extension String {
+    /// Split the string at each ";", remove any quoting.
+    /// 
+    /// The trouble is if there's a
+    /// ";" inside something that's quoted. And we can escape the separator and
+    /// the quotes with a "\".
+    var httpHeaderParts: (String, [(String,String?)])? {
```

Comment:
definitely that structure here would be nice, I am not certain what this return value is representing structurally.